core/src/tile/tileManager.cpp

```diff
@@ -39,8 +40,18 @@ TileManager::~TileManager() {
     m_tileSets.clear();
 }
 
-void TileManager::addDataSource(std::shared_ptr<DataSource>&& dataSource) {
+int32_t TileManager::addDataSource(std::shared_ptr<DataSource>&& dataSource) {
     m_tileSets.push_back({++m_tileSetSerial, dataSource});
+    return m_tileSetSerial;
+}
+
+std::shared_ptr<ClientGeoJsonSource> TileManager::getClientSourceById(int32_t _id) {
+    for (auto& set : m_tileSets) {
+        if (set.id == _id) {
+            return std::dynamic_pointer_cast<ClientGeoJsonSource>(set.source);
```

Comment:
instead of creating another shared_ptr here, we can return a const reference of the shared DataSource ptr. Something like the following diff:

File: core/src/tangram.cpp:
```
 void addSourcePoint(int _sourceId, double* _coords) {
-    auto source = m_tileManager->getClientSourceById(_sourceId);
+    const auto& source = std::dynamic_pointer_cast<ClientGeoJsonSource>(m_tileManager->getClientSourceById(_sourceId));
 
 void addSourceLine(int _sourceId, double* _coords, int _lineLength) {
-    auto source = m_tileManager->getClientSourceById(_sourceId);
+    const auto& source = std::dynamic_pointer_cast<ClientGeoJsonSource>(m_tileManager->getClientSourceById(_sourceId));
 
 void addSourcePoly(int _sourceId, double* _coords, int* _ringLengths, int _rings) {
-    auto source = m_tileManager->getClientSourceById(_sourceId);
+    const auto& source = std::dynamic_pointer_cast<ClientGeoJsonSource>(m_tileManager->getClientSourceById(_sourceId));
 
 void addSourceGeoJSON(int _sourceId, const char* _data) {
-    auto source = m_tileManager->getClientSourceById(_sourceId);
+    const auto& source = std::dynamic_pointer_cast<ClientGeoJsonSource>(m_tileManager->getClientSourceById(_sourceId));
```

File: core/src/tile/tileManager.cpp
```
-std::shared_ptr<ClientGeoJsonSource> TileManager::getClientSourceById(int32_t _id) {
+const std::shared_ptr<DataSource>& TileManager::getClientSourceById(int32_t _id) {
     for (auto& set : m_tileSets) {
         if (set.id == _id) {
-            return std::dynamic_pointer_cast<ClientGeoJsonSource>(set.source);
+            return set.source;
         }
     }
     return nullptr;
```

File: core/src/tile/tileManager.cpp
``` 
-    std::shared_ptr<ClientGeoJsonSource> getClientSourceById(int32_t _id);
+    const std::shared_ptr<DataSource>& getClientSourceById(int32_t _id);
```